src/managers/balancer.py

```diff
@@ -80,3 +181,114 @@ def create_internal_topics(self) -> None:
                     ],
                 )
                 logger.info(f"Created topic {topic}")
+
+    def rebalance(self, mode: str, dryrun: bool = True) -> requests.Response:
+        """Triggers a full Kafka cluster partition rebalance."""
+        rebalance_request = self.cruise_control.post(
+            "rebalance", dryrun=dryrun
+        )  # FIXME: will need updating to include add/remove_broker when it works
+        user_task_id = rebalance_request.headers.get("User-Task-ID", "")
+
+        # block entire charm event handling while rebalance in progress
```

Comment:
**thought**: I have a couple of thoughts / questions:
- How do you feel about adding a new maintenance status to be displayed while rebalance is ongoing?
- When you say _block entire charm event handling while rebalance in progress_, is it really the case? Shouldn't we shunt all events listeners based on `self.cruise_control.executing` to prevent config changes and stuff like that? 

My main concern would be that, even for a small deployment such as the one you demo-ed, we end up with an action timeout. The block is only temporary

Reply:
1. I think that would be an excellent idea! That said, we're choosing to completely block charm execution during the Action event execution. As statuses are set mid-execution (I actually misunderstood and thought it was at the end), a status would be a good suggestion, in addition to the logs I have. Will add to the PR now!

2. You're right that there are pros and cons to hard-blocking during a rebalance, VS what you suggested. It was my first idea honestly, but I spoke to the SQL teams about how they were doing backups (also a potentially long-running Action), and they do the same here, letting incoming events queue while the Action event handling completes. I think the approach taken in this PR is a bit easier to handle, as we don't have to be aware of what we can/can't block all the time

On your last point, Actions are async. The client makes a request, and then waits for a response some period of time before timing out. The charm itself handles the request until completion, it doesn't time out.
In theory, we could do `juju run balancer/leader rebalance mode=full dryrun=false --wait 100m` to extend that timeout to 100 minutes if we wanted to.

Reply:
Thank you for the clarification, I agree that the approach chosen here is the right way to go